### DISCUSSION TITLE
Get the context in a custom view

### DISCUSSION THREAD

**Discussion by lippoliv** [inventree/InvenTree#5625] [Q&A]
Hi folks,
I followed https://docs.inventree.org/en/latest/extend/plugins/navigation/ to add a new navigation item but I don't know how to get a proper context (spoiler: its not `request.context`). I thought it doesn't matter right now because I just want a text input, but on form submit (POC implementation) I am not logged in any longer.

# Screenshots

<img width="1551" alt="image" src="https://github.com/inventree/InvenTree/assets/2561859/3c653f2a-c4b7-40d1-91d9-7744be662287">

<img width="931" alt="image" src="https://github.com/inventree/InvenTree/assets/2561859/2a8a8176-88b0-44be-890a-55000a3e6952">

# POC Code

```python
from plugin import InvenTreePlugin
from plugin.mixins import SettingsMixin, NavigationMixin, UrlsMixin
from plugin.helpers import render_template

from django.http import HttpResponse
from django.conf.urls import url

import logging

logger = logging.getLogger('inventree')


class WooCommerceIntegrationPlugin(InvenTreePlugin, SettingsMixin,
                                   NavigationMixin, UrlsMixin):
    NAME = "WooCommerceIntegrationPlugin"
    SLUG = "mjwoocommerce"
    TITLE = "wooCommerce Import"

    # metadata
    AUTHOR = "Oliver Lippert"
    DESCRIPTION = "Import orders from wooCommerce"
    VERSION = "0.0.0"
    WEBSITE = "https://plugin.de/"
    LICENSE = "MIT"

    SETTINGS = {
        'API_URL': {
            'name': 'Website URL',
            'description': 'Your websites URL',
            'default': 'http://remote.url',
        },
        'API_KEY': {
            'name': 'Consumer Key',
            'description': '',
            'default': '',
            'required': True,
        },
        'API_SECRET': {
            'name': 'Consumer Secret',
            'description': '',
            'default': '',
            'required': True,
        },
    }

    NAVIGATION_TAB_NAME = "wooCommerce"
    NAVIGATION_TAB_ICON = 'fas fa-plus'

    NAVIGATION = [
        {
            'name': 'Import',
            'link': 'plugin:mjwoocommerce:import-form'
        },
    ]

    def setup_urls(self):
        return [
            url('show-import/', self.import_view, name='import-form'),
            url('run-import/', self.import_run, name='import-run'),
        ]

    def import_view(self, request):
        return HttpResponse(
            render_template(self, 'mjwoocommerce/import-form.html',{})
        )

    def import_run(self, request):
        return HttpResponse(
            request
        )
```

# Goal

In `import_view` I want to render a simple form with just an text input. On submit it should be handed over to `import_run` where I execute some code (or enqueue a background job, idk right now). That's it for the POC.

**Comment by lippoliv** [comment]
I switched my form from POST to GET and used `path` instead of `url` for registering `import-run`, now at least I continue to be logged in.

Still everything looks odd, but I can work on the POC....

**Comment by SchrodingersGat** [comment]
@lippoliv did you get any further with this? When you say "context" - do you mean within the template itself, for the custom rendering?

**Reply by lippoliv** [reply]
Haven't had time for deeper analysis here. I need support or I cannot move on.
Have a look on the UI screenshot. All menu items are missing. Here's the template code:

```txt
{% extends 'page_base.html' %}

{% block heading %}
  Bestellung importieren
{% endblock %}

{% block actions %}

{% endblock %}

{% block page_info %}
  hello there
{% endblock %}
```

I guess the menu is missing because the context is empty

```python
    def import_view(self, request):
        return HttpResponse(
            render_template(self, 'mjwoocommerce/import-form.html',{})
        )
```

The question is, where I get the "correct" context from?

**Reply by SchrodingersGat** [reply]
Ok, I think I get it. From `templates/navbar.html` I think that the header tabs are not displaying because the `roles` context var is not present:

https://github.com/inventree/InvenTree/blob/6f2dca729d213f2c18af75d8d4d8827126bf5f0e/InvenTree/templates/navbar.html#L21

This might be a pretty complex one - all our rendered views use django's class-based-views system which has a number of ways of passing in context. If you are directly calling `render_template` (without passing context) then you're going to have a bad time :)

For example, user role information is passed in as a global context variable here:

https://github.com/inventree/InvenTree/blob/6f2dca729d213f2c18af75d8d4d8827126bf5f0e/InvenTree/InvenTree/context.py#L61

And each InvenTree view has (potentially) custom context data defined as part of the class:

https://github.com/inventree/InvenTree/blob/6f2dca729d213f2c18af75d8d4d8827126bf5f0e/InvenTree/part/views.py#L307

So rendering a raw context response would be insufficient here. 

I would suggest that your path forward should be to instantiate a view class (inherited from existing classes much like ours do) and then pass the request through to the view. This *should* expose the rendering to all the available context...
